docs/cluster/networking.md

```diff
@@ -0,0 +1,142 @@
+running Docker and expect kubernetes to work.  You must ensure that the
+fundamental requirements are met.
+
+This model is not only less complex overall, but it is principally compatible
+with the desire for Kubernetes to enable low-friction migration from VMs.  If
+your job previously ran in a VM, your VM had an IP and could talk to other VMs
+in your project.  This is the same basic model.
+
+Until now this document has talked about containers.  In reality, Kubernetes
+applies IP addresses at the `Pod` scope - containers within a `Pod` share their
+network namespaces - including their IP address.  This means that containers
+within a `Pod` can all reach each other’s ports on `localhost`.  This does imply
+that containers within a `Pod` must coordinate port usage, but this is no
+different that processes in a VM.  We call this the "IP-per-pod" model.  It is
+possible to request host ports, but this is reduced to a very niche operation.
+
+## How to acheive this
+
+There are a number of ways that this network model can be implemented.  This
+document is not an exhaustive study of the various methods, but hopefully serves
+as an introduction to various technologies and serves as a jumping-off point.
+If some techniques become vastly preferable to others, we might detail them more
+here.
+
+### Google Compute Engine
+
+For the Google Compute Engine cluster configuration scripts, we use [advanced
+routing](https://developers.google.com/compute/docs/networking#routing) to
+assign each VM a subnet (default is /24 - 254 IPs).  Any traffic bound for that
+subnet will be routed directly to the VM by the GCE network fabric.  This is in
+addition to the "main" IP address assigned to the VM, which is NAT'ed for
+outbound internet access.  A linux bridge (called `cbr0`) is configured to exist
+on that subnet, and is passed to docker's `--bridge` flag.
+
+We start Docker with:
+
+```
+    DOCKER_OPTS="--bridge cbr0 --iptables=false --ip-masq=false"
+```
+
+We set up this bridge on each node with SaltStack, in
+[container_bridge.py](../cluster/saltbase/salt/_states/container_bridge.py).
+
+```
+cbr0:
+  container_bridge.ensure:
+    - cidr: {{ grains['cbr-cidr'] }}
+    - mtu: 1460
+```
+
+GCE itself does not know anything about these IPs, though, so it will not NAT
+them for outbound internet traffic.  To achieve that we us an iptables rule to
+masquerade traffic that is bound for IPs outside teh GCE project network
+(10.0.0.0/8).
+
+```
+iptables -t nat -A POSTROUTING ! -d 10.0.0.0/8 -o eth0 -j MASQUERADE
+```
+
+Lastly we enable IP forwarding in the kernel:
+
+```
+sysctl net.ipv4.ip_forward=1
+```
+
+The result of all this is that Docker allocates `Pod` IPs from the specified
+subnet, those IPs are all routable within the GCE project network, and
+containers can reach the internet.
+
+### L2 networks and linux bridging
+
+If you have a "dumb" L2 network you should be able to do something similar to
```

Comment:
RE: the new directory - I could see a definition of what makes a minimum Node setup going in that directory explaining the prerequisites?

Sent from my iPhone

> On Mar 5, 2015, at 8:52 PM, David Oppenheimer <notifications@github.com> wrote:
> 
> In docs/cluster/networking.md:
> 
> > +iptables -t nat -A POSTROUTING ! -d 10.0.0.0/8 -o eth0 -j MASQUERADE
> > +```
> > +
> > +Lastly we enable IP forwarding in the kernel:
> > +
> > +```
> > +sysctl net.ipv4.ip_forward=1
> > +```
> > +
> > +The result of all this is that Docker allocates `Pod` IPs from the specified
> > +subnet, those IPs are all routable within the GCE project network, and
> > +containers can reach the internet.
> > +
> > +### L2 networks and linux bridging
> > +
> > +If you have a "dumb" L2 network you should be able to do something similar to
> s/network/network, e.g. on a bare-metal cluster/
> 
> —
> Reply to this email directly or view it on GitHub.
>